### DISCUSSION TITLE
Host Avalonia control inside NSView on MacOS

### DISCUSSION THREAD

**Discussion by pauldumais** [AvaloniaUI/Avalonia#15719] [General]
Hello,

We would like to convert our large dotnet apps (winforms on Windows and Xamarin everwhere else) to Avalonia.  We have successfully started converting our  Windows app built with WinForms by using the WinFormsAvaloniaControlHost control inside our WinForms views, this works great so far.  This will allow us to over time incrementally convert the app from WinForms to Avalonia.

We also have a MacOS app built with Xamarin and Cocoa.  I'm first trying to embed an Avalonia view inside a Xamarin C# NSView similar to the WinFormsAvaloniaControlHost method, is this possible? I see in the native code folder an AvnView maybe we can expose this somehow to the C# layer?

I see a mention on this thread about accessing the NSWindow and @maxkatz6 mentioning that it might be possible to embed inside an NSView, we are willing to make the changes if needed to the native side to support this:
https://github.com/AvaloniaUI/Avalonia/discussions/15555#discussioncomment-9266757

Thanks,
Paul

**Comment by Gillibald** [comment]
I am currently refactoring the MacOS windowing implementation. It might be possible to implement a TopLevel that just holds a NSView. That could be embedded into some external NSView.

**Reply by pauldumais** [reply]
That's great, let me know if there is a branch I can try or if you would like some help, we are happy to assist.
Paul

**Reply by rbnpontes** [reply]
Hi, is there any update on this?

**Reply by pauldumais** [reply]
Yes this was added in 11.2.0

https://github.com/AvaloniaUI/Avalonia/pull/15932
